# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
df = pd.read_csv('../../data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [32]:
df.sort_values(by = 'Unit price', ascending = False).head(5)

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [34]:
filter_electronic_category = (df['Product line'] == 'Electronic accessories')
df[filter_electronic_category].head()

Invoice ID Branch       City Customer type  Gender  \
1   226-31-3081      C  Naypyitaw        Normal  Female   
5   699-14-3026      C  Naypyitaw        Normal    Male   
6   355-53-5943      A     Yangon        Member  Female   
11  529-56-3974      B   Mandalay        Member    Male   
12  365-64-0515      A     Yangon        Normal  Female   

              Product line  Unit price  Quantity   Tax 5%     Total  \
1   Electronic accessories       15.28         5   3.8200   80.2200   
5   Electronic accessories       85.39         7  29.8865  627.6165   
6   Electronic accessories       68.84         6  20.6520  433.6920   
11  Electronic accessories       25.51         4   5.1020  107.1420   
12  Electronic accessories       46.95         5  11.7375  246.4875   

         Date   Time  Payment    cogs  gross margin percentage  gross income  \
1    3/8/2019  10:29     Cash   76.40                 4.761905        3.8200   
5   3/25/2019  18:30  Ewallet  597.73                 4.761905       29.8865   
6   2/25/2019  14:36  Ewallet  413.04                 4.761905       20.6520   
11   3/9/2019  17:03     Cash  102.04                 4.761905        5.1020   
12  2/12/2019  10:25  Ewallet  234.75                 4.761905       11.7375   

    Rating  
1      9.6  
5      4.1  
6      5.8  
11     6.8  
12     7.1

In [9]:
df[filter_electronic_category].sort_values(by = ['Unit price', 'Quantity'], ascending = [False, False]).head(1)

Invoice ID Branch      City Customer type  Gender  \
209  817-69-8206      B  Mandalay        Normal  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
209  Electronic accessories       99.73         9  44.8785  942.4485   

         Date   Time      Payment    cogs  gross margin percentage  \
209  3/2/2019  19:42  Credit card  897.57                 4.761905   

     gross income  Rating  
209       44.8785     6.5

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [92]:
df['Rating Category'] = df['Rating']

In [94]:
def category_mapping(val):
    if val >= 8:
        return 'High'
    elif 8 > val >= 6:
        return 'Medium'
    else:
        return 'Low'

In [96]:
df['Rating Category'] = df.Rating.apply(category_mapping)
df[['Rating', 'Rating Category']].head(10)

Rating Rating Category
0     9.1            High
1     9.6            High
2     7.4          Medium
3     8.4            High
4     5.3             Low
5     4.1             Low
6     5.8             Low
7     8.0            High
8     7.2          Medium
9     5.9             Low

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [98]:
df['Rating Category'].value_counts(normalize=True)

Rating Category
Medium    0.356
High      0.329
Low       0.315
Name: proportion, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [100]:
time = '11:28'

In [116]:
def get_hour(time_str):
    return int(time_str.split(':')[0])

In [118]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [139]:
df['Hour'] = df['Time'].apply(get_hour)

In [187]:
df[['Hour', 'Time']].head(7)

Hour   Time
0    13  13:08
1    10  10:29
2    13  13:23
3    20  20:33
4    10  10:37
5    18  18:30
6    14  14:36

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [185]:
df.sort_values(by = ['Quantity', 'Hour'], ascending = [False, False]).head(10)

Invoice ID Branch       City Customer type  Gender  \
118  642-32-2990      A     Yangon        Normal  Female   
267  571-94-0759      B   Mandalay        Member  Female   
314  227-07-4446      C  Naypyitaw        Member  Female   
356  632-90-0281      C  Naypyitaw        Normal  Female   
457  533-33-5337      B   Mandalay        Normal    Male   
557  283-26-5248      C  Naypyitaw        Member  Female   
602  373-09-4567      C  Naypyitaw        Normal    Male   
14   829-34-3910      A     Yangon        Normal  Female   
91   225-32-0908      C  Naypyitaw        Normal  Female   
142  338-65-2210      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity  Tax 5%     Total  \
118      Food and beverages       10.96        10   5.480   115.080   
267      Food and beverages       74.60        10  37.300   783.300   
314  Electronic accessories       78.13        10  39.065   820.365   
356     Fashion accessories       37.55        10  18.775   394.275   
457  Electronic accessories       79.39        10  39.695   833.595   
557      Food and beverages       98.52        10  49.260  1034.460   
602      Food and beverages       77.56        10  38.780   814.380   
14        Health and beauty       71.38        10  35.690   749.490   
91        Sports and travel       44.86        10  22.430   471.030   
142       Health and beauty       68.60        10  34.300   720.300   

          Date   Time      Payment   cogs  gross margin percentage  \
118   2/2/2019  20:48      Ewallet  109.6                 4.761905   
267   1/8/2019  20:55         Cash  746.0                 4.761905   
314  2/10/2019  20:51         Cash  781.3                 4.761905   
356   3/8/2019  20:01  Credit card  375.5                 4.761905   
457   2/7/2019  20:24         Cash  793.9                 4.761905   
557  1/30/2019  20:23      Ewallet  985.2                 4.761905   
602  3/14/2019  20:35      Ewallet  775.6                 4.761905   
14   3/29/2019  19:21         Cash  713.8                 4.761905   
91   1/26/2019  19:54      Ewallet  448.6                 4.761905   
142   2/5/2019  19:57         Cash  686.0                 4.761905   

     gross income  Rating Rating Category  Hour  
118         5.480     6.0          Medium    20  
267        37.300     9.5            High    20  
314        39.065     4.4             Low    20  
356        18.775     9.3            High    20  
457        39.695     6.2          Medium    20  
557        49.260     4.5             Low    20  
602        38.780     6.9          Medium    20  
14         35.690     5.7             Low    19  
91         22.430     8.2            High    19  
142        34.300     9.1            High    19

Найбільше покупок відбувається о 20:00. Це цілком логічно, адже більшість людей працює вдень, а їхній робочий день зазвичай закінчується близько 18–19 години. Саме тому покупки найчастіше здійснюються ввечері, після завершення роботи.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [214]:
df.groupby('City')['Total'].mean()

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [219]:
df.groupby('City')['Total'].agg(['size', 'max', 'min', 'mean'])

size      max      min        mean
City                                         
Mandalay    332  1022.49  18.6375  319.872506
Naypyitaw   328  1042.65  10.6785  337.099715
Yangon      340  1039.29  12.6945  312.354031

Найбільше покупок здійснено в Yangon - 340. Це може бути спричинено високим попитом на товари, велика частина населення платоспроможна у цьому місті та регулярно здійснює покупки. Mandalay і Naypyitaw мають приблизно однаковий обсяг покупок (332 і 328 відповідно), але трохи поступаються Yangon.

Найвища максимальна покупка — в Naypyitaw 1042.65. Це може свідчити про високий рівень доходу покупців або можуть бути також і високі ціни товарів.

Найнижча мінімальна покупка — також у Naypyitaw 10.68. Цей показник дає розуміння, що тут асортимент товарів з нижчими цінами є більш доступний для населення, можливо на товари найпершої потреби.

Найвища середня сума покупки — в Naypyitaw 337.10 свідчить про більшу готовність покупців витрачати більше коштів за раз.
Найвища середня сума покупки в Yangon. Можливо тут покупці купують більше на менші чеки.

Загальна рекомендація по Yangon збільшити асортимент товару, додати знижки, програму лояльності, продумати над залученням покупців, щоб вони здійснювали покупки на більшу суму. По магазинах у інших містах — підтримувати рівень цін на товари, а також створювати акції, знижки на товари більш дорожчого сегменту (Naypyitaw), вивчити які товари купують частіше та у які години і додавати відповідно до асортименту та робити пропозиції у певні години, що стимулювати збільшення середнього розміру покупки у Mandalay.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [226]:
df.to_csv('supermarket_sales_enriched.csv', index=False)

In [229]:
pd.read_csv('supermarket_sales_enriched.csv', nrows=5)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10